src/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/StartupCode/StartupCodeHelpers.cs

```diff
@@ -49,25 +51,75 @@ internal static unsafe void InitializeCommandLineArgs(int argc, byte** argv)
+        private static unsafe void InitializeGlobalTables()
         {
-            int length = 0;
-            IntPtr strEETypePtr = GetModuleSection((int)ModuleSectionIds.StringEETypePtr, out length);
-            Contract.Assert(length == IntPtr.Size);
-
-            IntPtr strTableStart = GetModuleSection((int)ModuleSectionIds.StringFixupStart, out length);
-            Contract.Assert(length % IntPtr.Size == 0);
-
-            IntPtr strTableEnd = (IntPtr)((byte*)strTableStart + length);
+            IntPtr* modulesSectionStart;
+            IntPtr* modulesSectionEnd;
+            GetModulesSection(out modulesSectionStart, out modulesSectionEnd);
+            
+            for (IntPtr* current = modulesSectionStart; current < modulesSectionEnd; current++)
+            {
+                IntPtr* moduleHeaderStart = (IntPtr*)*current;
+                
+                if (moduleHeaderStart != null)
+                {
+                    // Module headers are always emitted in pairs
+                    Contract.Assert(current + 1 < modulesSectionEnd);
+
+                    IntPtr* moduleHeaderEnd = (IntPtr*)*(current + 1);
+                    IntPtr pModuleInfo = CreateModuleInfo((IntPtr)moduleHeaderStart, (IntPtr)moduleHeaderEnd);
+
+                    // Configure the module indirection cell with the newly created ModuleInfo. This allows EETypes to find
+                    // their interface dispatch map tables.
+                    int length;
+                    IntPtr* pSection = (IntPtr*)GetModuleSection(pModuleInfo, ModuleHeaderSection.ModuleIndirectionCell, out length);
+                    *pSection = pModuleInfo;
+
+                    // Initialize strings if any are present
+                    IntPtr pStringSection = GetModuleSection(pModuleInfo, ModuleHeaderSection.StringTable, out length);
+                    if (pStringSection != IntPtr.Zero)
+                    {
+                        Contract.Assert(length % IntPtr.Size == 0);
+                        InitializeStringTable(pStringSection, length);
+                    }
+
+                    // Initialize statics if any are present
+                    IntPtr pStaticsSection = GetModuleSection(pModuleInfo, ModuleHeaderSection.GCStaticRegion, out length);
+                    if (pStaticsSection != IntPtr.Zero)
+                    {
+                        Contract.Assert(length % IntPtr.Size == 0);
+                        InitializeStatics(pStaticsSection, length);
+                    }
+
+                    // Run eager class constructors if any are present
+                    IntPtr pEagerClassConstructorSection = GetModuleSection(pModuleInfo, ModuleHeaderSection.EagerCctor, out length);
+                    if (pEagerClassConstructorSection != IntPtr.Zero)
+                    {
+                        Contract.Assert(length % IntPtr.Size == 0);
+                        RunEagerClassConstructors(pEagerClassConstructorSection, length);
```

Comment:
The eager class constructors should run only once declarative initialization of all modules was done. (Otherwise, the eager class constructor can call into a different module that was not initialized yet.)